osu.Game/Graphics/UserInterface/HistoryTextBox.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Framework.Input.Events;
+using osu.Game.Rulesets.Difficulty.Utils;
+using osuTK.Input;
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public class HistoryTextBox : FocusedTextBox
+    {
+        private readonly LimitedCapacityQueue<string> messageHistory;
+
+        public IEnumerable<string> GetMessageHistory()
+        {
+            if (HistoryCount == 0)
+                yield break;
+
+            for (int i = HistoryCount - 1; i >= 0; i--)
+                yield return messageHistory[i];
+        }
+
+        public int HistoryCount => messageHistory.Count;
+
+        private int selectedIndex;
+
+        private string originalMessage = string.Empty;
+        private bool everythingSelected;
+
+        public string GetOldMessage(int index) => messageHistory[HistoryCount - index - 1];
+
+        public HistoryTextBox(int capacity = 100)
+        {
+            messageHistory = new LimitedCapacityQueue<string>(capacity);
+
+            Current.ValueChanged += text =>
+            {
+                if (string.IsNullOrEmpty(text.NewValue) || everythingSelected)
+                {
+                    selectedIndex = HistoryCount;
```

Comment:
Can you explain what the `everythingSelected` special case is doing here? I'm not sure I follow. Doesn't seem to be covered by tests either.

Reply:
The special case is ensuring reset of history indexing when a user is overwriting the message to a new one, as it may be common amongst users to use Control + A (`PlatformAction.SelectAll`) to quickly start writing a new message.
A `TestResetIndex` is covering this scenario.